<!DOCTYPE html><html lang="en" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0"><title>Exception Handling in C++ | Ryan Zhao</title><meta name="author" content="Ryan Zhao"><meta name="copyright" content="Ryan Zhao"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="Exception Handling in C++In this article, we will cover exception handling in C++ in great detail. We will discuss the motivation behind using exception handling, the basics of how to use it, and some">
<meta property="og:type" content="article">
<meta property="og:title" content="Exception Handling in C++">
<meta property="og:url" content="http://example.com/2024/04/01/Exception%20Handling%20in%20C++/index.html">
<meta property="og:site_name" content="Ryan Zhao">
<meta property="og:description" content="Exception Handling in C++In this article, we will cover exception handling in C++ in great detail. We will discuss the motivation behind using exception handling, the basics of how to use it, and some">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="http://example.com/img/avatar.JPG">
<meta property="article:published_time" content="2024-04-01T03:10:49.785Z">
<meta property="article:modified_time" content="2023-04-10T10:01:50.150Z">
<meta property="article:author" content="Ryan Zhao">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://example.com/img/avatar.JPG"><link rel="shortcut icon" href="/img/icon.png"><link rel="canonical" href="http://example.com/2024/04/01/Exception%20Handling%20in%20C++/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","languages":{"hits_empty":"We didn't find any results for the search: ${query}"}},
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  }
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: 'Exception Handling in C++',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2023-04-10 18:01:50'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
    win.getCSS = (url,id = false) => new Promise((resolve, reject) => {
      const link = document.createElement('link')
      link.rel = 'stylesheet'
      link.href = url
      if (id) link.id = id
      link.onerror = reject
      link.onload = link.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        link.onload = link.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(link)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/img/avatar.JPG" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">16</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">0</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">4</div></a></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/background.png')"><nav id="nav"><span id="blog-info"><a href="/" title="Ryan Zhao"><img class="site-icon" src="/img/icon.png"/><span class="site-name">Ryan Zhao</span></a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search" href="javascript:void(0);"><i class="fas fa-search fa-fw"></i><span> Search</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page" href="javascript:void(0);"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">Exception Handling in C++</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2024-04-01T03:10:49.785Z" title="Created 2024-04-01 11:10:49">2024-04-01</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2023-04-10T10:01:50.150Z" title="Updated 2023-04-10 18:01:50">2023-04-10</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/C/">C++</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="Exception Handling in C++"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="Exception-Handling-in-C"><a href="#Exception-Handling-in-C" class="headerlink" title="Exception Handling in C++"></a>Exception Handling in C++</h1><p>In this article, we will cover exception handling in C++ in great detail. We will discuss the motivation behind using exception handling, the basics of how to use it, and some best practices. This article also includes examples to help illustrate the concepts.</p>
<h2 id="Table-of-Contents"><a href="#Table-of-Contents" class="headerlink" title="Table of Contents"></a>Table of Contents</h2><ol>
<li>Introduction</li>
<li>Basic Exception Handling<ol>
<li>Throwing an Exception</li>
<li>Catching an Exception</li>
</ol>
</li>
<li>Standard Exceptions</li>
<li>Custom Exceptions</li>
<li>Exception Propagation</li>
<li>Best Practices</li>
<li>Conclusion</li>
</ol>
<p><a name="introduction"></a></p>
<h2 id="1-Introduction"><a href="#1-Introduction" class="headerlink" title="1. Introduction"></a>1. Introduction</h2><p>In programming, exceptions are unexpected events that occur during the execution of a program. They often result from erroneous input, hardware failures, or other exceptional conditions. Exception handling is a mechanism that allows developers to manage such situations gracefully, rather than letting the program crash or produce incorrect results.</p>
<p>C++ provides built-in support for exception handling through the <code>try</code>, <code>catch</code>, and <code>throw</code> keywords. This allows for a more structured and robust approach to error handling, making it easier to write reliable code.</p>
<p><a name="basic-exception-handling"></a></p>
<h2 id="2-Basic-Exception-Handling"><a href="#2-Basic-Exception-Handling" class="headerlink" title="2. Basic Exception Handling"></a>2. Basic Exception Handling</h2><p><a name="throwing-an-exception"></a></p>
<h3 id="2-1-Throwing-an-Exception"><a href="#2-1-Throwing-an-Exception" class="headerlink" title="2.1. Throwing an Exception"></a>2.1. Throwing an Exception</h3><p>In C++, you can throw an exception using the <code>throw</code> keyword followed by an expression. This expression represents the exception object and can be of any data type, including built-in types (e.g., <code>int</code>, <code>float</code>, <code>char</code>), objects, or pointers.</p>
<p>Here is an example of throwing an exception when a function is called with an invalid argument:</p>
<figure class="highlight c++"><table><tr><td class="code"><pre><span class="line"><span class="function"><span class="type">void</span> <span class="title">validateAge</span><span class="params">(<span class="type">int</span> age)</span> </span>&#123;</span><br><span class="line">    <span class="keyword">if</span> (age &lt; <span class="number">0</span>) &#123;</span><br><span class="line">        <span class="keyword">throw</span> <span class="string">&quot;Invalid age: Age cannot be negative.&quot;</span>;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><a name="catching-an-exception"></a></p>
<h3 id="2-2-Catching-an-Exception"><a href="#2-2-Catching-an-Exception" class="headerlink" title="2.2. Catching an Exception"></a>2.2. Catching an Exception</h3><p>To catch an exception, you need to enclose the code that may throw an exception within a <code>try</code> block. Then, you can specify one or more <code>catch</code> blocks to handle the exception(s). Each <code>catch</code> block must specify the type of exception it can handle.</p>
<p>Here’s an example of how to catch an exception:</p>
<figure class="highlight c++"><table><tr><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;iostream&gt;</span></span></span><br><span class="line"></span><br><span class="line"><span class="function"><span class="type">int</span> <span class="title">main</span><span class="params">()</span> </span>&#123;</span><br><span class="line">    <span class="type">int</span> age = <span class="number">-1</span>;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="built_in">validateAge</span>(age);</span><br><span class="line">    &#125; <span class="built_in">catch</span> (<span class="type">const</span> <span class="type">char</span>* errorMessage) &#123;</span><br><span class="line">        std::cerr &lt;&lt; <span class="string">&quot;Caught exception: &quot;</span> &lt;&lt; errorMessage &lt;&lt; std::endl;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> <span class="number">0</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>In this example, the <code>main</code> function calls the <code>validateAge</code> function within a <code>try</code> block. If <code>validateAge</code> throws an exception, the <code>catch</code> block following the <code>try</code> block will handle it. The <code>catch</code> block specifies that it can handle exceptions of type <code>const char*</code>, which matches the type of exception thrown by <code>validateAge</code>.</p>
<p><a name="standard-exceptions"></a></p>
<h2 id="3-Standard-Exceptions"><a href="#3-Standard-Exceptions" class="headerlink" title="3. Standard Exceptions"></a>3. Standard Exceptions</h2><p>C++ provides a set of standard exceptions defined in the <code>&lt;stdexcept&gt;</code> header. These exceptions are derived from the <code>std::exception</code> class and are designed to cover common error situations.</p>
<p>Some common standard exceptions include:</p>
<ul>
<li><code>std::invalid_argument</code></li>
<li><code>std::out_of_range</code></li>
<li><code>std::runtime_error</code></li>
<li><code>std::overflow_error</code></li>
<li><code>std::underflow_error</code></li>
</ul>
<p>Here’s an example ofusing a standard exception:</p>
<figure class="highlight c++"><table><tr><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;iostream&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;stdexcept&gt;</span></span></span><br><span class="line"></span><br><span class="line"><span class="function"><span class="type">void</span> <span class="title">validateAge</span><span class="params">(<span class="type">int</span> age)</span> </span>&#123;</span><br><span class="line">    <span class="keyword">if</span> (age &lt; <span class="number">0</span>) &#123;</span><br><span class="line">        <span class="keyword">throw</span> std::<span class="built_in">invalid_argument</span>(<span class="string">&quot;Invalid age: Age cannot be negative.&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="type">int</span> <span class="title">main</span><span class="params">()</span> </span>&#123;</span><br><span class="line">    <span class="type">int</span> age = <span class="number">-1</span>;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="built_in">validateAge</span>(age);</span><br><span class="line">    &#125; <span class="built_in">catch</span> (<span class="type">const</span> std::invalid_argument&amp; e) &#123;</span><br><span class="line">        std::cerr &lt;&lt; <span class="string">&quot;Caught exception: &quot;</span> &lt;&lt; e.<span class="built_in">what</span>() &lt;&lt; std::endl;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> <span class="number">0</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>In this example, we replaced the <code>const char*</code> exception with the <code>std::invalid_argument</code> exception. When catching the exception, we specify the <code>std::invalid_argument</code> type in the <code>catch</code> block and use the <code>what()</code> member function to get the error message.</p>
<p><a name="custom-exceptions"></a></p>
<h2 id="4-Custom-Exceptions"><a href="#4-Custom-Exceptions" class="headerlink" title="4. Custom Exceptions"></a>4. Custom Exceptions</h2><p>Sometimes, the standard exceptions may not cover all error situations in your application. In such cases, you can create custom exception classes by inheriting from the <code>std::exception</code> class or any of its derived classes.</p>
<p>Here’s an example of a custom exception class:</p>
<figure class="highlight c++"><table><tr><td class="code"><pre><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;stdexcept&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="keyword">include</span> <span class="string">&lt;string&gt;</span></span></span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">InvalidAgeException</span> : <span class="keyword">public</span> std::invalid_argument &#123;</span><br><span class="line"><span class="keyword">public</span>:</span><br><span class="line">    <span class="function"><span class="keyword">explicit</span> <span class="title">InvalidAgeException</span><span class="params">(<span class="type">const</span> std::string&amp; message)</span> : std::invalid_argument(message) &#123;</span>&#125;</span><br><span class="line">&#125;;</span><br></pre></td></tr></table></figure>

<p>To use this custom exception class, you can throw and catch it just like any other exception:</p>
<figure class="highlight c++"><table><tr><td class="code"><pre><span class="line"><span class="function"><span class="type">void</span> <span class="title">validateAge</span><span class="params">(<span class="type">int</span> age)</span> </span>&#123;</span><br><span class="line">    <span class="keyword">if</span> (age &lt; <span class="number">0</span>) &#123;</span><br><span class="line">        <span class="keyword">throw</span> <span class="built_in">InvalidAgeException</span>(<span class="string">&quot;Invalid age: Age cannot be negative.&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="type">int</span> <span class="title">main</span><span class="params">()</span> </span>&#123;</span><br><span class="line">    <span class="type">int</span> age = <span class="number">-1</span>;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="built_in">validateAge</span>(age);</span><br><span class="line">    &#125; <span class="built_in">catch</span> (<span class="type">const</span> InvalidAgeException&amp; e) &#123;</span><br><span class="line">        std::cerr &lt;&lt; <span class="string">&quot;Caught exception: &quot;</span> &lt;&lt; e.<span class="built_in">what</span>() &lt;&lt; std::endl;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> <span class="number">0</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><a name="exception-propagation"></a></p>
<h2 id="5-Exception-Propagation"><a href="#5-Exception-Propagation" class="headerlink" title="5. Exception Propagation"></a>5. Exception Propagation</h2><p>Exceptions can propagate through function calls. If a function does not catch an exception, it will be propagated up the call stack until it is caught by a suitable <code>catch</code> block or reaches the <code>main</code> function. If an exception reaches the <code>main</code> function without being caught, the program will terminate and the unhandled exception message will be displayed.</p>
<p>Here’s an example of exception propagation:</p>
<figure class="highlight c++"><table><tr><td class="code"><pre><span class="line"><span class="function"><span class="type">void</span> <span class="title">processData</span><span class="params">(<span class="type">int</span> age)</span> </span>&#123;</span><br><span class="line">    <span class="built_in">validateAge</span>(age);</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="type">int</span> <span class="title">main</span><span class="params">()</span> </span>&#123;</span><br><span class="line">    <span class="type">int</span> age = <span class="number">-1</span>;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="built_in">processData</span>(age);</span><br><span class="line">    &#125; <span class="built_in">catch</span> (<span class="type">const</span> InvalidAgeException&amp; e) &#123;</span><br><span class="line">        std::cerr &lt;&lt; <span class="string">&quot;Caught exception: &quot;</span> &lt;&lt; e.<span class="built_in">what</span>() &lt;&lt; std::endl;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> <span class="number">0</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>In this example, the <code>processData</code> function calls <code>validateAge</code>, but does not catch the exception. The exception propagates up to the <code>main</code> function, where it is caught and handled.</p>
<p><a name="best-practices"></a></p>
<h2 id="6-Best-Practices"><a href="#6-Best-Practices" class="headerlink" title="6. Best Practices"></a>6. Best Practices</h2><p>Here are some best practices for exception handling in C++:</p>
<ol>
<li>Use exceptions for error handling, not for regular control flow.</li>
<li>Use standard exceptions whenever possible, and create custom exceptions when necessary.</li>
<li>Catch exceptions by reference, not by value.</li>
<li>Don’t catch exceptions that you cannot handle. Let them propagate up the call stack to be handled at the appropriate level.</li>
<li>Be mindful of resource management (e.g., memory, file handles) when dealing with exceptions. Use RAII (Resource Acquisition Is Initialization) to manage resources.</li>
<li>Document the exceptions that your functions can throw, so that other developers know how to handle them.</li>
</ol>
<p><a name="conclusion"></a></p>
<h2 id="7-Conclusion"><a href="#7-Conclusion" class="headerlink" title="7. Conclusion"></a>7. Conclusion</h2><p>In this article, we covered exception handling in C++ in great detail. We discussed the motivation behind using exception handling, the basics of how to use it, and some best practices. We explored how to throw and catch exceptions, use standard exceptions, create custom exceptions, and handle exception propagation.</p>
<p>By following the best practices and understanding the concepts discussed in this article, you can write more robust and reliable C++ code. Exception handling allows you to manage unexpected events in a structured way, making it easier to handle errors and edge cases in your applications.</p>
<p>Remember to use exceptions for error handling and not for regular control flow, and always document the exceptions your functions may throw. By following these guidelines and using the examples provided as a reference, you’ll be well-equipped to handle exceptions in your C++ projects.</p>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="http://example.com">Ryan Zhao</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="http://example.com/2024/04/01/Exception%20Handling%20in%20C++/">http://example.com/2024/04/01/Exception%20Handling%20in%20C++/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"></div><div class="post_share"><div class="social-share" data-image="/img/avatar.JPG" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2024/04/01/File%20IO%20in%20C++/" title="File I/O in C++"><div class="cover" style="background: var(--default-bg-color)"></div><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">File I/O in C++</div></div></a></div><div class="next-post pull-right"><a href="/2024/04/01/C++%20Beginner's%20Guide/" title="C++ Beginner's Guide"><div class="cover" style="background: var(--default-bg-color)"></div><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">C++ Beginner's Guide</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/img/avatar.JPG" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">Ryan Zhao</div><div class="author-info__description">CS for everything</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">16</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">0</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">4</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/ryanzhao1226"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="https://github.com/ryanzhao1226" target="_blank" title="Github"><i class="fab fa-github"></i></a><a class="social-icon" href="https://twitter.com/ryan_zhao1226" target="_blank" title="Twitter"><i class="fab fa-twitter"></i></a><a class="social-icon" href="https://www.facebook.com/ryan.zhao.5496?mibextid=LQQJ4dhttps://www.facebook.com/ryan.zhao.5496?mibextid=LQQJ4d" target="_blank" title="Facebook"><i class="fab fa-facebook"></i></a><a class="social-icon" href="mailto:ryanzhao1226@gmail.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>Announcement</span></div><div class="announcement_content">Welcome to Ryan's blog! If you have any question, please contant me by E-mail!</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#Exception-Handling-in-C"><span class="toc-number">1.</span> <span class="toc-text">Exception Handling in C++</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#Table-of-Contents"><span class="toc-number">1.1.</span> <span class="toc-text">Table of Contents</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#1-Introduction"><span class="toc-number">1.2.</span> <span class="toc-text">1. Introduction</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#2-Basic-Exception-Handling"><span class="toc-number">1.3.</span> <span class="toc-text">2. Basic Exception Handling</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#2-1-Throwing-an-Exception"><span class="toc-number">1.3.1.</span> <span class="toc-text">2.1. Throwing an Exception</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#2-2-Catching-an-Exception"><span class="toc-number">1.3.2.</span> <span class="toc-text">2.2. Catching an Exception</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#3-Standard-Exceptions"><span class="toc-number">1.4.</span> <span class="toc-text">3. Standard Exceptions</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#4-Custom-Exceptions"><span class="toc-number">1.5.</span> <span class="toc-text">4. Custom Exceptions</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#5-Exception-Propagation"><span class="toc-number">1.6.</span> <span class="toc-text">5. Exception Propagation</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#6-Best-Practices"><span class="toc-number">1.7.</span> <span class="toc-text">6. Best Practices</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#7-Conclusion"><span class="toc-number">1.8.</span> <span class="toc-text">7. Conclusion</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2024/04/01/test/" title="test">test</a><time datetime="2024-04-01T05:27:10.332Z" title="Created 2024-04-01 13:27:10">2024-04-01</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2024/04/01/School%20Management%20System%20Development(C++)/" title="School Management System Development(C++)">School Management System Development(C++)</a><time datetime="2024-04-01T03:10:49.796Z" title="Created 2024-04-01 11:10:49">2024-04-01</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2024/04/01/Templates%20in%20C++/" title="Templates in C++">Templates in C++</a><time datetime="2024-04-01T03:10:49.796Z" title="Created 2024-04-01 11:10:49">2024-04-01</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2024/04/01/The%20Syntax%20in%20Java/" title="The Syntax in Java">The Syntax in Java</a><time datetime="2024-04-01T03:10:49.796Z" title="Created 2024-04-01 11:10:49">2024-04-01</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2024/04/01/The%20Syntax%20of%20C++/" title="The Syntax in C++">The Syntax in C++</a><time datetime="2024-04-01T03:10:49.796Z" title="Created 2024-04-01 11:10:49">2024-04-01</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2020 - 2024 By Ryan Zhao</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Switch Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.min.js"></script><div class="js-pjax"></div><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">Search</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  Loading the Database</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="Search for Posts" type="text"/></div></div><hr/><div id="local-search-results"></div></div></div><div id="search-mask"></div><script src="/js/search/local-search.js"></script></div></body></html>